src/request_queue.js

```diff
@@ -488,48 +511,54 @@ export class RequestQueue {
-                if (prevLimit >= REQUEST_QUEUE_HEAD_MAX_LIMIT) {
-                    log.warning(`'RequestQueue: We've reached the maximum number of requests in progress: ${REQUEST_QUEUE_HEAD_MAX_LIMIT}.'`);
-                }
-
-                // If checkModifiedAt=true then we must ensure that queueModifiedAt is older than
-                // queryStartedAt for at least API_PROCESSED_REQUESTS_DELAY_MILLIS.
-                const shouldRepeatForConsistency = checkModifiedAt && (queryStartedAt - queueModifiedAt < API_PROCESSED_REQUESTS_DELAY_MILLIS);
-
-                // If both are false then head is consistent and we may exit.
-                if (!shouldRepeatWithHigherLimit && !shouldRepeatForConsistency) return true;
+        const { queueModifiedAt, wasLimitReached, prevLimit, queryStartedAt, hadMultipleClients } = await this.queryQueueHeadPromise;
 
-                // If we are querying for consistency then we limit the number of queries to MAX_QUERIES_FOR_CONSISTENCY.
-                // If this is reached then we return false so that empty() and finished() returns possibly false negative.
-                if (!shouldRepeatWithHigherLimit && iteration > MAX_QUERIES_FOR_CONSISTENCY) return false;
+        // If queue is still empty then one of the following holds:
+        // - the other calls waiting for this promise already consumed all the returned requests
+        // - the limit was too low and contained only requests in progress
+        // - the whole queue was processed and we are done
 
-                const nextLimit = shouldRepeatWithHigherLimit
-                    ? prevLimit * 1.5
-                    : prevLimit;
+        // If limit was not reached in the call then there are no more requests to be returned.
+        const shouldRepeatWithHigherLimit = this.queueHeadDict.length() === 0 && wasLimitReached && prevLimit < REQUEST_QUEUE_HEAD_MAX_LIMIT;
+        if (prevLimit >= REQUEST_QUEUE_HEAD_MAX_LIMIT) {
+            log.warning(`RequestQueue: We've reached the maximum number of requests in progress: ${REQUEST_QUEUE_HEAD_MAX_LIMIT}.`);
+        }
 
-                const delayMillis = shouldRepeatForConsistency
-                    ? API_PROCESSED_REQUESTS_DELAY_MILLIS
-                    : 0;
+        // If checkModifiedAt=true then we must ensure that either:
+        // - queueModifiedAt is older than queryStartedAt for at least API_PROCESSED_REQUESTS_DELAY_MILLIS
+        // - hadMultipleClients=false and this.assumedTotalCount=this.assumedHandledCount
+        const isDatabaseConsistent = queryStartedAt - queueModifiedAt >= API_PROCESSED_REQUESTS_DELAY_MILLIS;
+        const isLocallyConsistent = !hadMultipleClients && this.assumedTotalCount === this.assumedHandledCount;
+        const shouldRepeatForConsistency = checkModifiedAt && (!isDatabaseConsistent && !isLocallyConsistent);
```

Comment:
Shouldn't this be `(!isDatabaseConsistent || !isLocallyConsistent)` ?

Reply:
BTW the code and variables here are quite hard to read, isn't there a way to make it more clear? (I don't know)